### PR TITLE
feat: support common modules

### PR DESCRIPTION
**Pull request by henworth**
# Pull Request

## Related Github Issues

- #155 
- #176
- #196
- #199  

## Description

This adds support for the "common module" pattern that was recently added to the [Terragrunt example repo](https://github.com/gruntwork-io/terragrunt-infrastructure-live-example). In addition it also automatically adds the included path as a dependency without requiring any `extra_atlantis_dependencies` blocks.

There is an important caveat to this approach however, _every_ include block is added as an implicit dependency, including any parent module. This means that irrespective of the state of  `--ignore-parent-terragrunt` the parent `terragrunt.hcl` will be included as there is no _easy_ way differentiate between includes to find the "real" parent. For that reason I'm marking this as a draft as we will need something akin to the idea discussed in #155 in order to properly support that functionality. 

I played around with the idea of adding a `is_parent_terragrunt_file` flag, but couldn't figure out the logic for it. That's the reason the `--include-parent-changes` cli argument still exists below, looks like I missed it in the change set.

This also adds a new functional cli argument `--create-parent-project` as I see a potential to _want_ to include the parent as a dependency but not create a project for it. So these two arguments are the two components of those changes, with `--ignore-parent-terragrunt` disabling both. These should provide enough flexibility to cover most uses of parent modules.

Lastly, there are two minor changes to how the error handling works. First, the printing of the usage/help is now suppressed when exiting with an error. This was something that kept coming up during debugging and it made things a bit easier to view. Also, this removes the second printing of any error, again because this was cleaner during debugging. 

## Security Implications

- _[none]_

## System Availability

- _[none]_

### CODE REVIEW

**Comment by dmattia** [comment]
This looks really interesting @henworth, thank you so much for this. Are you able to go through and update the existing tests by chance to include the new outputs? It would be useful for me to review to go through example by example and see what all the net effects are on the e2e tests

**Comment by henworth** [comment]
Absolutely I can do that. I just wanted to point it out before doing so.

File: cmd/generate.go [hunk 1]
```diff
@@ -106,6 +106,30 @@ func lookupProjectHcl(m map[string][]string, value string) (key string) {
 	return key
 }
 
+// func stringInSlice(a string, list []string) bool {
```
Comment (dmattia) [inline cmd/generate.go:109]: nit: remove

File: cmd/generate.go [hunk 2]
```diff
@@ -106,6 +106,30 @@ func lookupProjectHcl(m map[string][]string, value string) (key string) {
...
+// 	for _, b := range list {
+// 		if b == a {
+// 			return true
+// 		}
+// 	}
+// 	return false
+// }
+
+func sliceUnion(a, b []string) []string {
```
Comment (dmattia) [inline cmd/generate.go:118]: nit: add a super brief comment here for what the function does

**Comment by dmattia** [comment]
@henworth I just went through and reviewed the code, and it's wonderful, thank you so much for this contribution.

> every include block is added as an implicit dependency

I think this is sensible. It's definitely a change from how things have worked up until now with this tool, but I think with the evolution of terragrunt having multiple includes, this tradeoff is definitely necessary.

Once the tests pass here I will gladly merge + release 👍

**Comment by henworth** [comment]
@dmattia I've updated all the test golden outputs, and they all pass on my local dev. I'll take a quick pass at addressing the other comments you have. Thanks!

File: cmd/generate.go (hunk 1 above)
Comment (henworth) [reply cmd/generate.go]: Removed

File: cmd/generate.go (hunk 2 above)
Comment (henworth) [reply cmd/generate.go:110]: Added

**Comment by henworth** [comment]
I'm not sure why the tests are failing. In my forked repo the task was successful: https://github.com/henworth/terragrunt-atlantis-config/actions/runs/1901169810

Looks like it is failing on the `TestNotIgnoringParentTerragrunt` test: https://github.com/transcend-io/terragrunt-atlantis-config/runs/5340383108?check_suite_focus=true#step:6:31, but locally I can't reproduce it.

**Comment by smitthakkar96** [comment]
@henworth did you get chance to come back to this? I badly need this feature :P 

Hoping to see this releasing soon :)

**Comment by smitthakkar96** [comment]
I tried on my machine to reproduce however I am unable to as well, @dmattia can we please rerun the tests. Maybe a glitch or race condition that is not reproducible (if it doesn't work, retry :P), however it is better to find out

**Comment by henworth** [comment]
I've suspected something similar. I've been trying to debug it locally to find some strange edge case, but I can't find anything. Also considered the possibility of the version of Go itself causing the issue, but the test matrix is not consistent on which version will fail. 

Yesterday I ran the test on a loop for about 2 hours and there were only two failures in about 14,000 iterations. I'm honestly stumped right now.

**Comment by henworth** [comment]
And just like that... all the tests pass this time. ha

In regards to versions, I think it would be a good idea to drop tests for older non-supported versions of Go (1.13 and 1.14) and bump the "main" version to 1.16 to keep in step with Terragrunt. Atlantis is at 1.17 which I've added as a test version. Version 1.18 is also on the horizon which will cause 1.15 to drop out of support, so I think that could possibly be dropped as well.

**Review by smitthakkar96 — APPROVED** [review]

**Review by dmattia — APPROVED** [review]
Looks great, apologies for the test flakes, this is the first I've seen them 🤔

**Comment by smitthakkar96** [comment]
@dmattia did you create a release for this?
